src/main/scala/org/ergoplatform/ErgoBoxCandidate.scala

```diff
@@ -1,54 +1,68 @@
 package org.ergoplatform
 
 import com.google.common.primitives.Longs
+import org.bouncycastle.jcajce.provider.digest.SHA1.Digest
 import org.ergoplatform.ErgoBox._
+import scorex.crypto.encode.Base16
 import scorex.crypto.hash.Digest32
-import sigmastate.{SBoolean, SType}
-import sigmastate.Values.{ByteArrayConstant, Value, EvaluatedValue, LongConstant}
-import sigmastate.serialization.Serializer.{Position, Consumed}
-import sigmastate.serialization.{ValueSerializer, Serializer}
+import sigmastate._
+import sigmastate.Values.{ByteArrayConstant, ConcreteCollection, EvaluatedValue, Idn, LongConstant, Tuple, Value}
+import sigmastate.serialization.Serializer.{Consumed, Position}
+import sigmastate.serialization.{Serializer, ValueSerializer}
 import sigmastate.utxo.CostTable.Cost
 
 import scala.annotation.tailrec
+import scala.runtime.ScalaRunTime
 import scala.util.Try
 
 class ErgoBoxCandidate(val value: Long,
                        val proposition: Value[SBoolean.type],
-                       val additionalRegisters: Map[NonMandatoryIdentifier, _ <: EvaluatedValue[_ <: SType]] = Map()) {
+                       val additionalTokens: Seq[(TokenId, Long)] = Seq(),
+                       val additionalRegisters: Map[NonMandatoryRegisterId, _ <: EvaluatedValue[_ <: SType]] = Map()) {
 
   lazy val cost = (bytesWithNoRef.length / 1024 + 1) * Cost.BoxPerKilobyte
 
   val propositionBytes: Array[Byte] = proposition.bytes
 
   lazy val bytesWithNoRef: Array[Byte] = ErgoBoxCandidate.serializer.toBytes(this)
 
-  def toBox(txId: Array[Byte], boxId: Short) = ErgoBox(value, proposition, additionalRegisters, txId, boxId)
+  def toBox(txId: Array[Byte], boxId: Short) =
+    ErgoBox(value, proposition, additionalTokens, additionalRegisters, txId, boxId)
 
-  def get(identifier: RegisterIdentifier): Option[Value[SType]] = {
+  def get(identifier: RegisterId): Option[Value[SType]] = {
     identifier match {
-      case R0 => Some(LongConstant(value))
-      case R1 => Some(ByteArrayConstant(propositionBytes))
-      case R2 => None
-      case n: NonMandatoryIdentifier => additionalRegisters.get(n)
+      case ValueRegId => Some(LongConstant(value))
+      case ScriptRegId => Some(ByteArrayConstant(propositionBytes))
+      case TokensRegId =>
+        val tokenTuples = additionalTokens.map { case (id, amount) =>
+          Tuple(ByteArrayConstant(id), LongConstant(amount))
+        }.toIndexedSeq
+        Some(ConcreteCollection(tokenTuples, STuple(SCollection[SByte.type], SLong)))
```

Comment:
@kushti just FYI, there is a shortcut `val SByteArray = SCollection(SByte)`. It's systematic usage will make code shorter, and also more efficient, because equals uses eq to check for reference equality. See https://stackoverflow.com/questions/4526706/what-code-is-generated-for-an-equals-hashcode-method-of-a-case-class

Reply:
fixed